hooks/image/70-flatpak-appstream-catalog

```diff
@@ -0,0 +1,88 @@
+
+import glob
+import logging
+import os
+import subprocess
+import sys
+from xml.etree import ElementTree
+
+logging.basicConfig(
+    level=logging.INFO,
+    format='+ %(asctime)s %(levelname)s %(name)s: %(message)s',
+    datefmt='%H:%M:%S'
+)
+logger = logging.getLogger(os.path.basename(__file__))
+
+def main():
+    root = os.path.join(os.environ['OSTREE_VAR'], 'lib/flatpak/exports')
+
+    # Find the metainfo files to include in the AppStream.
+    ids = []
+    for metainfo in glob.glob(os.path.join(root, 'share/metainfo/*.xml')):
+        # Everything here should end in .metainfo.xml or .appdata.xml, but
+        # let's be sure.
+        if not (metainfo.endswith('.metainfo.xml') or
+                metainfo.endswith('.appdata.xml')):
+            raise Exception(f'Unexpected metainfo file {metainfo}')
+
+        # It's kind of bad to reject these files here because they're obviously
+        # valid enough for Flathub to have published them. Reasons that appstream-util
+        # 0.7.18-1+deb11u1 considers appdata to be invalid include:
+        #
+        # - Using a <url type=''> attribute from a more recent version of the spec
+        #   (dozens of apps, including GIMP)
+        #
+        # - A <li> element contains a nested <em> element (Pods; OK this is not a
+        #   preinstalled app, but still)
+        #
+        # - <release> elements are not in order (FileRoller)
+        #
+        # None of these are reasons to exclude the app from the catalogue but there is
+        # no way to tell the tool to ignore them: if we pass these files to
+        # appstream-compose it will fail.
+        #
+        # TODO: Switch to appstreamcli validate-relax once we are on Bookworm and can
+        # use appstreamcli compose.
+        cmd = ('appstream-util', 'validate-relax', '--nonet', metainfo)
+        proc = subprocess.run(cmd)
+        if proc.returncode != 0:
+            logger.warning('Skipping %s, considered invalid by appstream-util', metainfo)
+            continue
+
+        # Convert the filename to an "app-id" as expected by
+        # appstream-compose. This isn't necessarily the app ID but the stem
+        # of the metainfo XML file without .metainfo.xml or .appdata.xml.
+        # Most of the time these are the same.
+        metainfo_id = os.path.basename(metainfo)
+        if metainfo_id.endswith('.metainfo.xml'):
+            metainfo_id = metainfo_id.rpartition('.metainfo.xml')[0]
+        else:
+            metainfo_id = metainfo_id.rpartition('.appdata.xml')[0]
+        logger.info('Including metainfo ID %s', metainfo_id)
+        ids.append(metainfo_id)
+
+    # Exit if there aren't any desktop metainfo files.
+    if len(ids) == 0:
+        logger.info('No metainfo files')
+        return
+
+    # Compose the AppStream XML.
+    # TODO: Switch to appstreamcli compose once we are on Bookworm
+    cmd = [
+        'appstream-compose',
+        f'--origin={os.environ["EIB_OUTVERSION"]}',
```

Comment:
This causes the filename to be something like `eoscustom-master-amd64-amd64.231110-173131.base.xml.gz`. [According to the spec](https://www.freedesktop.org/software/appstream/docs/chap-CatalogData.html#sect-AppStream-XML) this is an appropriate filename:

> The XML files must have a unique name, which is usually the distribution's name and version, combined with the name of the repository/origin. For example in Debian 8 (Jessie), the filename for the main repository component would be `debian-jessie-main.xml.gz`. For Fedora 20 (Heisenbug) updates it would be `fedora-20-updates.xml.gz`. 3rd-party repositories use a vendor name and repository-name combination, for example Ubuntu PPAs might get `ppa-ubuntu12.04-username-foobar.xml`.

We might like to put `.catalog` or `.appstream-catalog` in the filename I guess?

Reply:
I'd like that. Like `.appstream.xml.gz` or something. To me that's clear enough that it's a catalog since you'd use `.metainfo.xml` if it was a single component. But I don't feel super strongly either way. Only that a bare `.xml` file will seem ambiguous in the context of image build assets and not "some place where you put appstream catalogs".